### PR TITLE
meta(typescript): convert utils to typescript

### PR DESCRIPTION
**Pull request by sdepold**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [ ] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Convert utils to typescript.

### CODE REVIEW

File: lib/utils/class-to-invokable.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
  * Wraps a constructor to not need the `new` keyword using a proxy.
  * Only used for data types.
  *
- * @param {Function} Class The class instance to wrap as invocable.
+ * @param {ProxyConstructor} Class The class instance to wrap as invocable.
  * @returns {Proxy} Wrapped class instance.
  * @private
  */
-function classToInvokable(Class) {
+export function classToInvokable(Class: any): Proxify {
```
Comment (sdepold) [inline lib/utils/class-to-invokable.ts:11]: This is a bit too hardcore for my understanding. Theoretically, `Class` should be of type `ProxyConstructor` but that causes problems with the spread operator: 

```A spread argument must either have a tuple type or be passed to a rest parameter.```

**Comment by sdepold** [comment]
Hmmmm. I'm trying to fix the stub but it somehow doesn't work

File: lib/utils/class-to-invokable.ts (hunk 1 above)
Comment (AllAwesome497) [reply lib/utils/class-to-invokable.ts:11]: Something like this should work:

```ts
export function classToInvokable<
  T extends (new () => any) // the () might not be necessary, can't remember. 
>(Class: T): T extends new (...infer Args) => infer Instance 
  ? (...args: Args) => Instance & (new (...args: Args) => Instance) & (Args['length'] extends 0 ? Instance : {})
   : never // this won't ever if  T is a constructor, its just here since we need an `extends` to use `infer`.
 { /* function body */}
```

(haven't verified, may need a bit of tweaking).

File: lib/utils/join-sql-fragments.ts [hunk 2]
```diff
@@ -1,54 +1,55 @@
...
  * Otherwise, if a non-string, non-falsy value is present, a TypeError will be thrown.
- * 
+ *
  * @returns {string} The joined string.
- * 
+ *
  * @private
  */
-function joinSQLFragments(array) {
+export function joinSQLFragments(array: any[]): string {
```
Comment (AllAwesome497) [inline lib/utils/join-sql-fragments.ts:52]: This should be valid (unsure if ts v3.9 will support it):
```ts
type Falsy = 0 | '' | null | undefined | false;
type SQLFragment = string | Falsy | SQLFragment[];
```


w/ the function signature being:
```suggestion
export function joinSQLFragments(array: SQLFragment[]): string {
```

Though this requires a bit of tweaking to get the function working properly (iirc the ts assertions will break a lil - I also converted this file to TS locally and I mostly rewrote the function + optimized it a bit).

File: lib/utils/join-sql-fragments.ts [hunk 3]
```diff
@@ -80,4 +85,8 @@ function joinSQLFragments(array) {
 
   return singleSpaceJoinHelper(array);
 }
-exports.joinSQLFragments = joinSQLFragments;
+
+type JoinSQLFragmentsError = TypeError & {
+  args?: any[];
+  fragment?: any;
+};
```
Comment (AllAwesome497) [inline lib/utils/join-sql-fragments.ts:92]: nit: can you make this extend `Error` for a standard JS error?

ie:
```ts
class JoinSQLFragmentsError extends TypeError {
  args: SQLFragment[];
  fragment: any; // iirc this error is only used when we get an invalid fragment.

  constructor(args: SQLFragment[], fragment: any, message: string) {
    super(message);
    this.args = args;
    this.fragment = fragment;
    // This is good to do explicitly for some bundlers / minnification. 
    this.name = 'JoinSQLFragmentsError';
  }
}
```

File: lib/utils/class-to-invokable.ts (hunk 1 above)
Comment (sdepold) [reply lib/utils/class-to-invokable.ts:11]: what.. the... hell 😂

File: test/integration/dialects/abstract/connection-manager.test.js [hunk 4]
```diff
@@ -125,10 +142,15 @@ describe(Support.getTestDialectTeaser('Connection Manager'), () => {
     };
 
     await connectionManager.getConnection(queryOptions);
-    chai.expect(deprecationStub).to.have.been.calledOnce;
+    chai.expect(stub).to.have.been.calledOnce;
+    chai
+      .expect(stub.getCalls()[0].args[0])
+      .to.contain(
+        'This database engine version is not supported, please update your database server.'
```
Comment (sdepold) [inline test/integration/dialects/abstract/connection-manager.test.js:149]: I changed this to this format now. The rest of the file is just formatting.

Instead of checking that the deprecation function was called, I'm checking one level further down where deprecate is calling process.emitWarning.

File: lib/generic/falsy.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export type Falsy = 0 | '' | null | undefined | false;
```
Comment (WikiRik) [inline lib/generic/falsy.ts:1]: Do we want to include -0, 0n and NaN as well? As per https://developer.mozilla.org/en-US/docs/Glossary/Falsy

File: test/integration/dialects/abstract/connection-manager.test.js [hunk 6]
```diff
@@ -1,14 +1,12 @@
...
-  sinon = require('sinon'),
-  Config = require('../../../config/config'),
-  ConnectionManager = require('sequelize/lib/dialects/abstract/connection-manager'),
-  Pool = require('sequelize-pool').Pool;
-
+const chai = require('chai');
+const { unsupportedEngine } = require('../../../../lib/utils/deprecations');
+const Support = require('../../support');
+const sinon = require('sinon');
+const ConnectionManager = require('../../../../lib/dialects/abstract/connection-manager');
```
Comment (WikiRik) [inline test/integration/dialects/abstract/connection-manager.test.js:5]: Why was it an absolute path before and why did you change it to a relative path? (not knowing why it was this in the first place is a valid answer)

**Review by WikiRik — APPROVED** [review]
Don't know about the specific TypeScript questions you posted, but it looks good enough to be merged for me. I do have two questions, but they are non-blocking

File: lib/generic/falsy.ts (hunk 5 above)
Comment (sdepold) [reply lib/generic/falsy.ts]: Seems that NaN is actually no type (I think it is a number actually)

File: test/integration/dialects/abstract/connection-manager.test.js (hunk 6 above)
Comment (sdepold) [reply test/integration/dialects/abstract/connection-manager.test.js]: fixed

**Review by WikiRik — APPROVED** [review]
Looks good enough to be merged, restricting the Class and fragment is non-blocking and something for further improvements

File: lib/generic/falsy.ts (hunk 5 above)
Comment (AllAwesome497) [reply lib/generic/falsy.ts]: Yea `NaN` doesn't have a constant type in TS and is instead declared as a `number`, despite there being a float32 value specifically for NaN (unsure if JS actually implements this) :/.
